cmake/app/boilerplate.cmake

```diff
@@ -239,7 +220,83 @@ foreach(root ${BOARD_ROOT})
   if(BOARD_DIR AND NOT (${root} STREQUAL ${ZEPHYR_BASE}))
     set(USING_OUT_OF_TREE_BOARD 1)
   endif()
+endforeach()
+
+# Check that BOARD_REVISION has not changed.
+zephyr_check_cache(BOARD_REVISION)
+
+if(NOT ACTIVE_BOARD_REVISION)
+  file(GLOB revision_candidates LIST_DIRECTORIES false RELATIVE ${BOARD_DIR}
+       ${BOARD_DIR}/${BOARD}_[0-9]*.conf
+  )
+
+  foreach(candidate ${revision_candidates})
+    if(${candidate} MATCHES "${BOARD}_((0|[1-9]+)(_[0-9]+)?(_[0-9]+)?).conf")
+      string(REPLACE "_" "." FOUND_BOARD_REVISION ${CMAKE_MATCH_1})
+
+      if(DEFINED BOARD_REVISION)
+        if(${BOARD_REVISION} STREQUAL ${FOUND_BOARD_REVISION})
+          set(ACTIVE_BOARD_REVISION ${BOARD_REVISION} CACHE INTERNAL
+              "Active board revision")
+          break()
+        elseif(
+          (${BOARD_REVISION} VERSION_GREATER_EQUAL ${FOUND_BOARD_REVISION}) AND
+          (${FOUND_BOARD_REVISION} VERSION_GREATER_EQUAL "${ACTIVE_BOARD_REVISION}")
+        )
+          set(ACTIVE_BOARD_REVISION ${FOUND_BOARD_REVISION} CACHE INTERNAL
+              "Active board revision (Differs from user selected)")
```

Comment:
I don't think falling back like this is a good idea. If the user specifies a version and it cannot be matched exactly, I think that should be an error.

Otherwise, if I accidentally make a typo. I may get unexpected results. We can't trust users to read the CMake output in detail.

Is this feature truly necessary? If there is a hard requirement for it somewhere, maybe it can be turned on using an additional variable like `BOARD_REVISION_ALLOW_INEXACT=<TRUE|FALSE>` or so.

I also don't think this is really following semantic versioning. I think if the user requests board revision 2.0.0 but gets 1.2.6, then semver would say they are getting an incompatible version. If there is a major version change (major version 2 is not major version 1), then backward compatibility is explicitly not guaranteed.

So perhaps this should be limited to patch versions only, e.g. requesting 1.2.0 would match 1.2.7 (since, by semantic versioning, 1.2.7 is compatible with 1.2.0), but not 1.1.0 (since that is not compatible).

Reply:
> I don't think falling back like this is a good idea. If the user specifies a version and it cannot be matched exactly, I think that should be an error.
>
> Otherwise, if I accidentally make a typo. I may get unexpected results. We can't trust users to read the CMake output in detail.
> Is this feature truly necessary?

Thanks for some good thoughts.

The reason I chose to default to this behavior in this proposal is that companies may produce a series of boards with different revisions that requires no changes in Zephyr, as example:
Board versions:
- 0.1.0
- 0.2.0
- 0.3.0
- 0.4.0
- 1.0.0
- 1.1.0
- 1.2.0  <-- This have have changes that requires updated Zephyr Kconfig / DTS settings
- 1.3.0  

Now, in without the range possibility, then there would be a need for having:
- `<board>_0_1_0.conf`
- `<board>_0_2_0.conf`
- `<board>_0_3_0.conf`
- `<board>_0_4_0.conf`
- `<board>_1_0_0.conf`
- `<board>_1_2_0.conf`
- `<board>_1_3_0.conf`

and whenever a new board variant is created, then another file is needed, even if no changes are needed.
Board variants are not created / numbered with Zephyr in mind.
With the current proposal, there is only a need for adding the `<board>_1_2_0.conf`, where the changes are happening.

I wanted to make sure that users that have a board, let's say: nrf52840dk_nrf53840, rev 0.12.0, can simply type that revision and not having to care if that revision has an exact match.

Remember, this is a hw board, so the hw revision might actually change just because a new diode is added to the board, a component has changed from A->B, and thus board is in practice the same (== no changes in Zephyr).
But it could also be a SoC update. A SoC update would most likely be a major version change on the board, but it could be the SoC is software compatible (hence no Zephyr changes are needed).

Or the SoC change could be an completely new SoC (and thus probably a new board, as discussed here: https://github.com/zephyrproject-rtos/zephyr/pull/29990#issuecomment-726749897)

Looking at me desk, I have boards numbered 0.5.0, 0.7.0, 0.9.0, 0.12.0,  1.0.0, 1.2.1, Rev A, Rev B.
So being able to just type the revision and be sure Zephyr picks a compatible configuration, while at the same time allow a board maintainer to only update board files when there is an incompatible change in a new board revision was my goal.
(and make sure that a new board revision would not require yet an update if it could already be handle).

> If there is a hard requirement for it somewhere, maybe it can be turned on using an additional variable like BOARD_REVISION_ALLOW_INEXACT=<TRUE|FALSE> or so.

Well, so far I don't have any requirements, but I know board revision support was a wanted / needed feature.
But with the multitude of ways board manufactures define revisions, then there would not be one solution that would fit all, which is why I also allow for `revision.cmake`

There could be a `set(BOARD_REVISION_STRICT true)` if we believe there is a need, and board maintainer could set this in `revision.cmake`, but with current proposal, then it would be just as easy to write:
```
# revision.cmake
set(valid_revisions 0.1.0 0.5.0 1.0.0 1.2.0)

if(NOT (${BOARD_REVISION} IN_LIST valid_revisions))
     message(FATAL_ERROR "Invalid revision specified")
endif()

set(ACTIVE_BOARD_REVISION ${BOARD_REVISION} CACHE INTERNAL "Active board revision")
```

> So perhaps this should be limited to patch versions only, e.g. requesting 1.2.0 would match 1.2.7 (since, by semantic versioning, 1.2.7 is compatible with 1.2.0), but not 1.1.0 (since that is not compatible).

In software, that would be true, but now we are talking about hw, and hw designers may think of incompatibility in a different way, and jump the major version if they decide to use an updated version of the SoC (which could be software compatible).
They could of course also decide to completely change the SoC.

If looking at the PR description, the reel board revision jump from v1 --> v2 is still using same SoC:
https://github.com/zephyrproject-rtos/zephyr/blob/4d930057d6d667634752e3d0d0dbd677761714e2/boards/arm/reel_board/reel_board_defconfig#L3
https://github.com/zephyrproject-rtos/zephyr/blob/4d930057d6d667634752e3d0d0dbd677761714e2/boards/arm/reel_board/reel_board_v2_defconfig#L3

whereas the efm32pg_stk3402a / efm32pg_stk3402a_jg board is changing SoC:
https://github.com/zephyrproject-rtos/zephyr/blob/4d930057d6d667634752e3d0d0dbd677761714e2/boards/arm/efm32pg_stk3402a/efm32pg_stk3402a_defconfig#L3
https://github.com/zephyrproject-rtos/zephyr/blob/4d930057d6d667634752e3d0d0dbd677761714e2/boards/arm/efm32pg_stk3402a/efm32pg_stk3402a_jg_defconfig#L4

I see your point, and the strict use-case is indirectly supported with little code but of course a revision.cmake with just:
```
set(BOARD_REVISION_STRICT true)
```
would be even shorter than the example above.

Currently I would like to see this actively used, and then adjust if boards adopting to use board revision starts making everything strict.

Reply:
> Currently I would like to see this actively used, and then adjust if boards adopting to use board revision starts making everything strict.

I still really think that this (very strange and not at all "semantic") versioning scheme should be dropped as the default behavior, and boards should have to opt-in to it using a revision.cmake that explicitly asks for it.

Reply:
yes, currently i'm just waiting to see if more people have comments.

Reply:
> very strange and not at all "semantic"

just a small comment, "semantic" is no longer use, and the 'very strange' is similar to how CMake does things when it comes to `find_package()`.
If you do `find_package(<package> VERSION <version>)`, CMake will find the package that claims compatibility with `<version>`. So if you ask for version `1.3.0`, and only have `1.6.0` installed, and `1.6.0` claims compatibility, then you'll be using `1.6.0`.
https://cmake.org/cmake/help/latest/command/find_package.html?highlight=find_package#version-selection

Of course, it's the package that decides what compatible means (which is where we have `revision.cmake`).

Also note, if you do:
```
find_package(Python 3.11)
execute_process(COMMAND ${Python_EXECUTABLE} --version OUTPUT_VARIABLE version ERROR_VARIABLE version)
message("Python version: ${version}")
```
you'll see:
```
-- Could NOT find Python: Found unsuitable version "3.8.0", but required is at least "3.11" (found /usr/bin/python3)
Python found: /usr/bin/python3
Python version: Python 3.8.0
```
so it informs that it cannot find the requested, and then falls back to what is available.

But again, that's the `FindPython.cmake` package implementation that decides that behavior, not CMake itself.

Reply:
> just a small comment, "semantic" is no longer use,

The PR description says: "This PR introduces support for semantic versioning of boards."

> and the 'very strange' is similar to how CMake does things when it comes to find_package().
> [...] Of course, it's the package that decides what compatible means (which is where we have `revision.cmake`).

Right, which is the same was what I'm asking for. Every board should need a revision.cmake to decide what compatible means. Helpers for common cases can be brought in via include, function call, etc.

Reply:
> The PR description says: "This PR introduces support for semantic versioning of boards."

Fixed. I had updated the commit messages themselves, but forgotten the PR description.

> Right, which is the same was what I'm asking for. Every board should need a revision.cmake to decide what compatible means. 

I know. Before doing re-work (mostly on the docs), then I would just like some more feedback.
Especially whether Zephyr should have a default way of indicating what compatible means (so that not all boards need to do this with slight variations)

Reply:
@mbolivar-nordic code and documentation reworked.

A `revision.cmake` is now required for any board supporting the `-DBOARD=<board>@<revision>` feature.

A `board_check_revision` function is available to make it simpler for board maintainers to implement a revision scheme for their board.

The `board_check_revision` support the following revision formats:
- `LETTER`, revision is a single letter, A-Z.
- `MAJOR`, revision is a single digit, `x`.
- `MAJOR.MINOR`, revision is two digits, `x.y`.
- `MAJOR.MINOR.PATCH`, revision is three digits, `x.y.z`.

but a board maintainer is still free to implement any custom revision handling scheme directly in `revision.cmake`.